Migrates contacts add + edit

dianabarsan opened the pull request:
# Description

# Code review items

- Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- Tested: Unit and/or e2e where appropriate
- Internationalised: All user facing text
- Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

dianabarsan commented:
hey @njogz , mind having a look? 

Thanks for your time!

latin-panda commented on webapp/src/ts/modules/contacts/contacts-edit.component.ts:
```diff
@@ -0,0 +1,318 @@
+    let formId;
+    let titleKey;
+    const typeId = contact?.contact_type || contact?.type || this.routeSnapshot.params?.type;
+    return this.contactTypesService.get(typeId).then(type => {
```
```suggestion
    return this.contactTypesService
      .get(typeId)
      .then(type => {
```
Also, we can `await` so we can flat the statements below by removing the `then`

dianabarsan replied:
I don't like mixing async/await with promises in the same file, it feels messy. So I'm not switching to async/await because that would warrant a complete rewrite, which is risky.

latin-panda commented on webapp/src/ts/modules/contacts/contacts-edit.component.ts:
```diff
@@ -0,0 +1,318 @@
+      } else { // adding
+        this.contact = {
+          type: 'contact',
+          contact_type: this.routeSnapshot.params.type,
```
```suggestion
          contact_type: this.routeSnapshot.params?.type,
```

latin-panda commented on webapp/src/ts/modules/contacts/contacts-edit.component.ts:
```diff
@@ -0,0 +1,318 @@
+        this.contact = {
+          type: 'contact',
+          contact_type: this.routeSnapshot.params.type,
+          parent: this.routeSnapshot.params.parent_id
```
```suggestion
          parent: this.routeSnapshot.params?.parent_id
```

latin-panda commented on webapp/src/ts/services/contact-save.service.ts:
```diff
@@ -0,0 +1,191 @@
+    const preparedSiblings = [];
+    let promiseChain = Promise.resolve();
+
+    this.CONTACT_FIELD_NAMES.forEach(fieldName => {
```
This is a very long forEach callback, can we extract it in functions by concerns (if that makes sense here) ? :)

latin-panda commented on webapp/src/ts/services/contact-save.service.ts:
```diff
@@ -0,0 +1,191 @@
+      .then(() => docId ? this.dbService.get().get(docId) : null)
+      .then(original => {
+        const submitted = this.enketoTranslationService.contactRecordToJs(form.getDataStr({ irrelevant: false }));
+        if (original) {
```
This is like pre-preparing docs, is it possible to put this logic inside the `prepareSubmittedDocsForSave ` ? :)

latin-panda reviewed (commented):
Since I'm working today, I thought on helping with the review, I hope that's okay.
This is great, clean and easy to follow! I just left a couple of small comments but nothing big

dianabarsan commented:
Thanks a lot for the review @latin-panda . 
Besides the existing requests, I changed a couple more things:

1. I cleared a todo that I had forgotten about from tasks (to resolve actions of type "contact"). 
2. I fixed the "no-parent" contact create route. 
Unfortunately, I had to change the route we used in angularJs `/contacts//add/<type>` because the Angular router doesn't parse routes that have double slashes (https://stackoverflow.com/a/61094277), so now the route to add contacts without parent is: `/contacts/add/<type>`. 

Could you, please, have another look? Thanks again!

dianabarsan commented:
@latin-panda I just rebased to resolve conflicts and no other changes were made.

latin-panda reviewed (approved):
This is great! I reviewed, tested the branch and looks good. I don't have additional feedback :)
